Feat/add support for bitbucket sync

TheeMattOliver opened the pull request:
## tl;dr

Draft PR to assess the state of Bitbucket sync and token storage for Figma Tokens. Goal: bring Bitbucket sync for Figma Tokens up to speed with other git providers

## What's working

- authenticate and check correct permissions
- push custom commit message
- push commit data content as new commit to main branch
- push commit data content as new commit to designated branch
- creating pull requests

## What's left to do 

- pull data from Bitbucket repo by implementing the `read` method in `BitbucketTokenStorage.ts`
- create a tree structure from the contents of a Bitbucket directory to support multifile sync
- handle paginated lists of branches in `fetchBranches` in `BitbucketTokenStorage.ts`
- improve unit test coverage

## Files of interest:

remaining work to be done in this file &rarr; `src/storage/BitbucketTokenStorage.ts`

And for reference: 

`src/storage/RemoteTokenStorage.ts`
`src/app/store/remoteTokens.tsx`
`src/app/store/updateSources.tsx`
`src/store/bitbucket/bitbucket.tsx`

`src/app/components/SyncSettings.tsx`
`src/app/components/StorageItemForm/BitbucketForm`
`src/app/components/PushDialog`

## Note on the `read` method
Other git providers have wrapper methods that have provided a `createTree` method to create a tree structure from the contents of a directory. See [here](https://octokit.github.io/rest.js/v18#git-create-tree) and [here](https://docs.gitlab.com/ee/api/repositories.html), for example.

Neither the [batteries-included wrapper library](https://github.com/MunifTanjim/node-bitbucket) we're using nor the [raw Bitbucket API](https://developer.atlassian.com/cloud/bitbucket/rest/api-group-source/#api-repositories-workspace-repo-slug-src-commit-path-get) provide that.

I've tried to look for guidance in the Atlassian docs but all I can find is that [supposedly this endpoint supports recursion if you specify additionally the max_depth parameter and indicate how many levels you want to have.](https://developer.atlassian.com/cloud/bitbucket/rest/api-group-source/#api-repositories-workspace-repo-slug-src-commit-path-get) Haven't been able to try that. My feeling is that we may have to craft a recursive traversal method similar to `git ls-tree` to support `read` in `BitbucketTokenStorage`. I could be wrong about that and I am open to suggestions!

## Note on unit tests 
The methods that are implemented are working in the plug-in, but I think I'm doing something incorrectly in the way I'm mocking those functions or passing return values into the mocked functions in Jest. Open to suggestions for correctly mocking them to improve test coverage.

Those are in `/storage/___tests___/BitbucketTokenStorage.test.ts`.

LiamMartens commented:
The parts that are already implemented look pretty solid. If we want to put the UI part of this behind a feature flag all we need to do is the following:
- Add the flag variable in `flags.d.ts` just for the typings. We can add something like `bitbucketSync`
- In the `SyncSettings.tsx` file you can then use our wrapped `useFlags` hook (from `app/components/LaunchDarkly`) to check for the availability of the flag and hide/show the UI to select the new provider accordingly. Then we could be ok to merge this in (after some of the comments) and pick up the finalization of the new provider separately

CC @six7

six7 commented:
> The parts that are already implemented look pretty solid. If we want to put the UI part of this behind a feature flag all we need to do is the following:
> 
> * Add the flag variable in `flags.d.ts` just for the typings. We can add something like `bitbucketSync`
> * In the `SyncSettings.tsx` file you can then use our wrapped `useFlags` hook (from `app/components/LaunchDarkly`) to check for the availability of the flag and hide/show the UI to select the new provider accordingly. Then we could be ok to merge this in (after some of the comments) and pick up the finalization of the new provider separately
> 
> CC @six7

Sounds good. Let's do that. 

@TheeMattOliver Can you make sure to add those flags? Then we could merge this first version and later on expand on that. It's in a state where it's ready to be tested though, right? I think we'll have a large amount of testers for this as it's such a highly requested feature.

I've setup a `bitBucketSync` flag in Launch Darkly and enabled it for your user. Once you implemented it you can test it by removing your key then the flag should be false.

TheeMattOliver commented:
Thanks!

 > * In the `SyncSettings.tsx` file you can then use our wrapped `useFlags` hook (from `app/components/LaunchDarkly`) to check for the availability of the flag and hide/show the UI to select the new provider accordingly. Then we could be ok to merge this in (after some of the comments) and pick up the finalization of the new provider separately

I'm not sure I understand how to use this since I've never actually used feature flags myself 🙈 If you point me in the right direction I can read up on it though!

> @TheeMattOliver Can you make sure to add those flags types? It's called `bitBucketSync` now. 

Done, no problem

> It's in a state where it's ready to be tested though, right? I think we'll have a large amount of testers for this as it's such a highly requested feature.

Before you let people start mashing on it I think we should probably implement the `read` method

> I've setup a `bitBucketSync` flag in Launch Darkly and enabled it for your user. Once you implemented it you can test it by removing your key then the flag should be false.

six7 commented:
> Thanks!
> 
> > * In the `SyncSettings.tsx` file you can then use our wrapped `useFlags` hook (from `app/components/LaunchDarkly`) to check for the availability of the flag and hide/show the UI to select the new provider accordingly. Then we could be ok to merge this in (after some of the comments) and pick up the finalization of the new provider separately
> 
> I'm not sure I understand how to use this since I've never actually used feature flags myself 🙈 If you point me in the right direction I can read up on it though!

For example in `ThemeSelector.tsx` we check the presence of the `tokenThemes` flag. If true, show something, if not, don't.

It would be like this for the bitBucketSync flag:
```
import { useFlags } from '../LaunchDarkly';
const { bitBucketSync } = useFlags();

…

if (bitBucketSync) {
  // code when flag is true
}

```

TheeMattOliver commented:
Got it, thanks! Please hold, will update and push

six7 commented on src/utils/deepMerge.ts:
```diff
@@ -0,0 +1,62 @@
+/* ts-deep-merge, unlicensed, https://github.com/voodoocreation/ts-deepmerge/blob/master/src/index.ts */
```
Do we need this file? It's not used anywhere and it's unlicensed?

six7 reviewed (changes requested):
Can we remove the deepMerge file? Other than that I'd be happy to merge this in 👍

TheeMattOliver commented:
phew the commit history is a mess but this branch should be good to merge in now ✅

six7 commented on src/storage/BitbucketTokenStorage.ts:
```diff
@@ -0,0 +1,179 @@
+import { Bitbucket } from 'bitbucket';
+import { RemoteTokenStorageFile } from './RemoteTokenStorage';
+import {
+  GitMultiFileObject, GitSingleFileObject, GitStorageMetadata, GitTokenStorage,
+} from './GitTokenStorage';
```
```suggestion
import { GitStorageMetadata, GitTokenStorage } from './GitTokenStorage';
```

six7 commented on src/storage/BitbucketTokenStorage.ts:
```diff
@@ -0,0 +1,179 @@
+        workspace: this.owner,
+        repo_slug: this.repository,
+        ref: originRef,
+      });
```
```suggestion
      });
      console.log('originBranch', originBranch);
      console.log('newRef', newRef);
```

six7 replied:
need to do this as otherwise we get a typescript error because you have unused variables

six7 commented on src/storage/BitbucketTokenStorage.ts:
```diff
@@ -0,0 +1,179 @@
+        repo_slug: this.repository,
+        // path: this.path,
+        // ref: this.branch,
+      });
```
```suggestion
      });
      console.log('response', response);
```

six7 replied:
need to do this as otherwise we get a typescript error because you have unused variables

six7 commented on src/app/hooks/useIsGitMultiFileEnabled.ts:
```diff
@@ -6,10 +6,15 @@ import { StorageProviderType } from '@/constants/StorageProviderType';
-      && !api?.filePath?.endsWith('.json'),
-    )
-  ), [api]);
+  return useMemo(
+    () => Boolean(
+      multiFileSync
+          && (api?.provider === StorageProviderType.GITHUB
+            || api?.provider === StorageProviderType.GITLAB
+            || api?.provider === StorageProviderType.ADO
+            || api?.provider === StorageProviderType.BITBUCKET)
+          && !api?.filePath?.endsWith('.json'),
+    ),
+    [api, multiFileSync],
+  );
```
```suggestion
  return useMemo(
    () => Boolean((api?.provider === StorageProviderType.GITHUB
            || api?.provider === StorageProviderType.GITLAB
            || api?.provider === StorageProviderType.ADO
            || api?.provider === StorageProviderType.BITBUCKET)
          && !api?.filePath?.endsWith('.json')),
    [api],
  );
```

six7 commented on src/storage/__tests__/BitbucketTokenStorage.test.ts:
```diff
@@ -0,0 +1,203 @@
+import { TokenTypes } from '@/constants/TokenTypes';
+import { TokenSetStatus } from '@/constants/TokenSetStatus';
```
```suggestion
// import { TokenTypes } from '@/constants/TokenTypes';
// import { TokenSetStatus } from '@/constants/TokenSetStatus';
```

six7 commented on src/storage/__tests__/BitbucketTokenStorage.test.ts:
```diff
@@ -0,0 +1,203 @@
+
+  it('canWrite should return true if user has admin or write permissions', async () => {
+    mockGetAuthedUser.mockImplementationOnce(() => Promise.resolve({
+      currentUser: { data: { account_id: '' } },
```
```suggestion
      data: { account_id: '123' },
```

six7 replied:
seems like that data had the wrong structure

six7 commented on src/storage/__tests__/BitbucketTokenStorage.test.ts:
```diff
@@ -0,0 +1,203 @@
+        },
+      },
+    ]);
+
+    expect(mockCreateOrUpdateFiles).toBeCalledWith({
+      branch: 'main',
+      owner: 'MattOliver',
+      repo: 'figma-tokens-testing',
+      createBranch: false,
+      changes: [
+        {
+          message: 'Initial commit',
+          files: {
+            'data/tokens.json': JSON.stringify(
+              {
+                $themes: [
+                  {
+                    id: 'light',
+                    name: 'Light',
+                    selectedTokenSets: {
+                      global: TokenSetStatus.ENABLED,
+                    },
+                  },
+                ],
+                global: {
+                  red: {
+                    type: TokenTypes.COLOR,
+                    name: 'red',
+                    value: '#ff0000',
+                  },
+                },
+              },
+              null,
+              2,
+            ),
+          },
+        },
+      ],
+    });
+  });
```
```suggestion
  // it('should be able to write', async () => {
  //   mockListBranches.mockImplementationOnce(() => Promise.resolve({
  //     data: { values: [{ name: 'main' }] },
  //   }));

  //   mockCreateOrUpdateFiles.mockImplementationOnce(() => Promise.resolve({
  //     owner: 'MattOliver',
  //     repo: 'figma-tokens-testing',
  //     branch: 'main',
  //     changes: [
  //       {
  //         message: 'Initial commit',
  //         files: {
  //           'data/tokens.json':
  //               '{\n'
  //               + '  "$themes": [\n'
  //               + '    {\n'
  //               + '      "id": "light",\n'
  //               + '      "name": "Light",\n'
  //               + '      "selectedTokenSets": {\n'
  //               + '        "global": "enabled"\n'
  //               + '      }\n'
  //               + '    }\n'
  //               + '  ],\n'
  //               + '  "global": {\n'
  //               + '    "red": {\n'
  //               + '      "type": "color",\n'
  //               + '      "name": "red",\n'
  //               + '      "value": "#ff0000"\n'
  //               + '    }\n'
  //               + '  }\n'
  //               + '}',
  //         },
  //       },
  //     ],
  //   }));

  //   storageProvider.changePath('data/tokens.json');
  //   await storageProvider.write([
  //     {
  //       type: 'metadata',
  //       path: '$metadata.json',
  //       data: {},
  //     },
  //     {
  //       type: 'themes',
  //       path: '$themes.json',
  //       data: [
  //         {
  //           id: 'light',
  //           name: 'Light',
  //           selectedTokenSets: {
  //             global: TokenSetStatus.ENABLED,
  //           },
  //         },
  //       ],
  //     },
  //     {
  //       type: 'tokenSet',
  //       name: 'global',
  //       path: 'global.json',
  //       data: {
  //         red: {
  //           type: TokenTypes.COLOR,
  //           name: 'red',
  //           value: '#ff0000',
  //         },
  //       },
  //     },
  //   ], {
  //     commitMessage: '',
  //   });

  //   expect(mockCreateOrUpdateFiles).toBeCalledWith({
  //     branch: 'main',
  //     owner: 'MattOliver',
  //     repo: 'figma-tokens-testing',
  //     createBranch: false,
  //     changes: [
  //       {
  //         message: 'Initial commit',
  //         files: {
  //           'data/tokens.json': JSON.stringify(
  //             {
  //               $themes: [
  //                 {
  //                   id: 'light',
  //                   name: 'Light',
  //                   selectedTokenSets: {
  //                     global: TokenSetStatus.ENABLED,
  //                   },
  //                 },
  //               ],
  //               global: {
  //                 red: {
  //                   type: TokenTypes.COLOR,
  //                   name: 'red',
  //                   value: '#ff0000',
  //                 },
  //               },
  //             },
  //             null,
  //             2,
  //           ),
  //         },
  //       },
  //     ],
  //   });
  // });
```

six7 replied:
i had to comment that whole test out because it's failing

six7 commented on src/storage/__tests__/BitbucketTokenStorage.test.ts:
```diff
@@ -0,0 +1,203 @@
+        },
+      },
+    ]);
+
+    expect(mockCreateOrUpdateFiles).toBeCalledWith({
+      branch: 'main',
+      owner: 'MattOliver',
+      repo: 'figma-tokens-testing',
+      createBranch: false,
+      changes: [
+        {
+          message: 'Initial commit',
+          files: {
+            'data/tokens.json': JSON.stringify(
+              {
+                $themes: [
+                  {
+                    id: 'light',
+                    name: 'Light',
+                    selectedTokenSets: {
+                      global: TokenSetStatus.ENABLED,
+                    },
+                  },
+                ],
+                global: {
+                  red: {
+                    type: TokenTypes.COLOR,
+                    name: 'red',
+                    value: '#ff0000',
+                  },
+                },
+              },
+              null,
+              2,
+            ),
+          },
+        },
+      ],
+    });
+  });
```
```suggestion
  // it('should be able to write', async () => {
  //   mockListBranches.mockImplementationOnce(() => Promise.resolve({
  //     data: { values: [{ name: 'main' }] },
  //   }));

  //   mockCreateOrUpdateFiles.mockImplementationOnce(() => Promise.resolve({
  //     owner: 'MattOliver',
  //     repo: 'figma-tokens-testing',
  //     branch: 'main',
  //     changes: [
  //       {
  //         message: 'Initial commit',
  //         files: {
  //           'data/tokens.json':
  //               '{\n'
  //               + '  "$themes": [\n'
  //               + '    {\n'
  //               + '      "id": "light",\n'
  //               + '      "name": "Light",\n'
  //               + '      "selectedTokenSets": {\n'
  //               + '        "global": "enabled"\n'
  //               + '      }\n'
  //               + '    }\n'
  //               + '  ],\n'
  //               + '  "global": {\n'
  //               + '    "red": {\n'
  //               + '      "type": "color",\n'
  //               + '      "name": "red",\n'
  //               + '      "value": "#ff0000"\n'
  //               + '    }\n'
  //               + '  }\n'
  //               + '}',
  //         },
  //       },
  //     ],
  //   }));

  //   storageProvider.changePath('data/tokens.json');
  //   await storageProvider.write([
  //     {
  //       type: 'metadata',
  //       path: '$metadata.json',
  //       data: {},
  //     },
  //     {
  //       type: 'themes',
  //       path: '$themes.json',
  //       data: [
  //         {
  //           id: 'light',
  //           name: 'Light',
  //           selectedTokenSets: {
  //             global: TokenSetStatus.ENABLED,
  //           },
  //         },
  //       ],
  //     },
  //     {
  //       type: 'tokenSet',
  //       name: 'global',
  //       path: 'global.json',
  //       data: {
  //         red: {
  //           type: TokenTypes.COLOR,
  //           name: 'red',
  //           value: '#ff0000',
  //         },
  //       },
  //     },
  //   ], {
  //     commitMessage: '',
  //   });

  //   expect(mockCreateOrUpdateFiles).toBeCalledWith({
  //     branch: 'main',
  //     owner: 'MattOliver',
  //     repo: 'figma-tokens-testing',
  //     createBranch: false,
  //     changes: [
  //       {
  //         message: 'Initial commit',
  //         files: {
  //           'data/tokens.json': JSON.stringify(
  //             {
  //               $themes: [
  //                 {
  //                   id: 'light',
  //                   name: 'Light',
  //                   selectedTokenSets: {
  //                     global: TokenSetStatus.ENABLED,
  //                   },
  //                 },
  //               ],
  //               global: {
  //                 red: {
  //                   type: TokenTypes.COLOR,
  //                   name: 'red',
  //                   value: '#ff0000',
  //                 },
  //               },
  //             },
  //             null,
  //             2,
  //           ),
  //         },
  //       },
  //     ],
  //   });
  // });
```

six7 commented:
@TheeMattOliver I merged next in, fixed the merge conflicts and fixed some of your tests. You had a test that was testing that we can write, so I commented that out as it doesnt work yet. If possible apply those suggestions then we can merge.

TheeMattOliver commented:
> @TheeMattOliver I merged next in, fixed the merge conflicts and fixed some of your tests. You had a test that was testing that we can write, so I commented that out as it doesnt work yet. If possible apply those suggestions then we can merge.

Thank you! 

Committed those changes -- let me know if there's more to do!

six7 reviewed (approved):
Had another conflict, fixed it! Let's merge this in so we can continue 🎉
